src/Polly/Timeout/AsyncTimeoutEngine.cs

```diff
@@ -15,17 +15,28 @@ internal static class AsyncTimeoutEngine
         TimeSpan timeout = timeoutProvider(context);
 
         using var timeoutCancellationTokenSource = new CancellationTokenSource();
-        using var combinedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCancellationTokenSource.Token);
+        // Do not use a using here, the exception will exit the scope before we have time to
+        // notify the downstream of the cancellation.
+        var combinedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCancellationTokenSource.Token);
 
         Task<TResult> actionTask = null;
         CancellationToken combinedToken = combinedTokenSource.Token;
 
         try
         {
+
+            var actionTaskFn = async () =>
+            {
+                var result = await action(context, combinedToken).ConfigureAwait(continueOnCapturedContext);
+                // dispose of the token source after we've waited on the task to finish.
+                combinedTokenSource.Dispose();
+                return result;
+            };
```

Comment:
My concern here is that, unless I'm misunderstanding, this creates scenarios where the pessimistic timeout doesn't happen and an exception is thrown (i.e. any other type of exception) so the disposal is then left to the finalizer.

I can see [in the dotnet/runtime source code](https://github.com/dotnet/runtime/blob/4322fc9f5732fbbe0d95727cc75b439e66196906/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs#L884-L973) that things won't fire if the token source is disposed of more than once, but the registrations should still file when the scope exits and it's disposed of?

Reply:
Maybe just explicitly cancelling it via the source if it wasn't already in the `finally` block would resolve it?

Reply:
@martincostello -- I will `try` that and report back.

Reply:
That seems to fix it! Nice one, thanks!